lld/COFF/Symbols.cpp

```diff
@@ -112,20 +112,20 @@ DefinedImportThunk::DefinedImportThunk(COFFLinkerContext &ctx, StringRef name,
                                        ImportThunkChunk *chunk)
     : Defined(DefinedImportThunkKind, name), wrappedSym(s), data(chunk) {}
 
-Defined *Undefined::getWeakAlias() {
+Symbol *Undefined::getWeakAlias() {
   // A weak alias may be a weak alias to another symbol, so check recursively.
   DenseSet<Symbol *> weakChain;
   for (Symbol *a = weakAlias; a; a = cast<Undefined>(a)->weakAlias) {
```

Comment:
So... previously, this `cast<Undefined>(a)` was broken (invoking UB etc), if we had a weak alias which pointed at something which wasn't either `Undefined` or `Defined`?

Reply:
Yes.

The assumption that a symbol is either `Undefined` or `Defined` is true for some callers, but not all. See the other comment for details.